rust/azure_iot_operations_connector/src/file_mount_azure_device_registry.rs

```diff
@@ -1,4 +1,1110 @@
+                }
+                Ok(Self {
+                    device_name: device_name.to_string(),
+                    endpoint_name: endpoint_name.to_string(),
+                })
+            }
+            None => Err(Error(ErrorKind::ParseError(
+                "Failed to parse DeviceEndpointRef from string".to_string(),
+            ))),
+        }
+    }
+}
+
+impl TryFrom<&PathBuf> for DeviceEndpointRef {
+    type Error = Error;
+
+    fn try_from(value: &PathBuf) -> Result<Self, Self::Error> {
+        // TODO: Handle case where file name is not a file but a directory
+        // TODO: Handle case where file name has invalid UTF-8 characters (remove need for to_string_lossy)
+        value
+            .file_name()
+            .ok_or(ErrorKind::ParseError("File path ends in ..".to_string()))?
+            .to_string_lossy()
+            .to_string()
+            .try_into()
+    }
+}
+
+impl Display for DeviceEndpointRef {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}_{}", self.device_name, self.endpoint_name)
+    }
+}
+
+/// Represents an asset associated with a specific device and endpoint.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct AssetRef {
+    /// The name of the asset
+    pub name: String,
+    /// The name of the device
+    pub device_name: String,
+    /// The name of the endpoint
+    pub endpoint_name: String,
+}
+
+// ~~~~~~~~~~~~~~~~~ Observations ~~~~~~~~~~~~~~~~~~~~~
+
+/// Represents an observation for device endpoint creation events.
+#[allow(dead_code)]
+pub struct DeviceEndpointCreateObservation {
+    /// A file watcher used to monitor changes in the file mount.
+    debouncer: notify_debouncer_full::Debouncer<RecommendedWatcher, RecommendedCache>,
+    /// A channel for receiving notifications about device endpoint creation events.
+    create_device_rx: UnboundedReceiver<(DeviceEndpointRef, AssetCreateObservation)>,
+}
+
+impl DeviceEndpointCreateObservation {
+    /// Creates an instance of [`DeviceEndpointCreateObservation`] to observe device endpoint creation events.
+    ///
+    /// Returns Ok([`DeviceEndpointCreateObservation`]) if the observation is successfully created, otherwise returns an [`Error`].
+    ///
+    /// # Arguments
+    /// * `debounce_duration` - The duration to debounce incoming I/O events.
+    ///
+    /// # Errors
+    /// - [`Error`] of kind [`ErrorKind::EnvironmentVariableError`] if the environment variable was not able to be read.
+    /// - [`Error`] of kind [`ErrorKind::WatcherError`] if the watcher could not be created.
+    /// - [`Error`] of kind [`ErrorKind::IoError`] if there are issues accessing the file mount.
+    /// - [`Error`] of kind [`ErrorKind::ParseError`] if there are issues parsing the file names and content.
+    pub fn new(debounce_duration: Duration) -> Result<DeviceEndpointCreateObservation, Error> {
```

Comment:
maybe weird question, but would we want to use an options builder struct so this can have a default? If not, a recommended default might be good to have in the docs

Reply:
Good point, I added a default in the example so customers can choose to copy that one, I'll add a comment here too